### PR TITLE
added firebase auth and made prisma client globally usable

### PR DESCRIPTION
**Pull request by gnorbsl**

### CODE REVIEW

File: src/helper/authMiddleware.ts [hunk 1]
```diff
@@ -1,16 +1,53 @@
...
-import { Role, Roles } from '../entities/user';
-import { extractTokenFromAuthorization } from './jwt';
+import { Role } from '../entities/user';
+import prisma from '../db/prisma';
 
-export const authMiddleware = (
+admin.initializeApp();
+
+async function getAndSetUserRole(token: auth.DecodedIdToken) {
+  const user = await prisma.user.findUnique({
```
Comment (Enubia) [inline src/helper/authMiddleware.ts:12]: try-catch this please in case the DB is not reachable

File: src/helper/authMiddleware.ts [hunk 2]
```diff
@@ -1,16 +1,53 @@
...
+    where: {
+      email: token.email,
+    },
+  });
+
+  if (!user) {
+    throw new Error('CREATE NEW USER OR SOMETHING');
```
Comment (Enubia) [inline src/helper/authMiddleware.ts:19]: use our custom error class for this ;)

File: src/helper/authMiddleware.ts [hunk 3]
```diff
@@ -1,16 +1,53 @@
...
+      .auth()
+      .verifyIdToken(req.headers.authorization);
+
+    req.user = {
+      email: decodedToken.email,
+      role: decodedToken.role || (await getAndSetUserRole(decodedToken)),
+      uid: decodedToken.uid,
+    };
+  } catch (error) {
+    console.log(error);
```
Comment (Enubia) [inline src/helper/authMiddleware.ts:45]: logger.error instead of console

**Review by Enubia — CHANGES_REQUESTED** [review]
please also run the tests locally and fix the occurring errors

**Review by Enubia — CHANGES_REQUESTED** [review]

File: src/helper/authMiddleware.ts (hunk 1 above)
Comment (gnorbsl) [reply src/helper/authMiddleware.ts:12]: there is already a try catch in the calling function

File: src/db/prisma.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { PrismaClient } from '@prisma/client';
+
+// Prevent multiple instances of Prisma Client in development
+declare const global: NodeJS.Global & { prisma?: PrismaClient };
+
+const prisma = global.prisma || new PrismaClient();
+if (process.env.NODE_ENV === 'development') {
+  global.prisma = prisma;
```
[the remark this answers is not in the document]
Comment (gnorbsl) [reply src/db/prisma.ts:8]: how can I use apollo context in the auth middleware?

File: package.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
     "build": "tsc",
     "clean": "rimraf ./node_modules ./dist",
     "coverage": "jest --runInBand --coverage",
-    "dev": "nodemon --inspect=0.0.0.0:9229 -e ts,js -w src --exec node -r ./loadEnv.js -r ts-node/register src/",
+    "dev": "GOOGLE_APPLICATION_CREDENTIALS=donate-gifts.json nodemon --inspect=0.0.0.0:9229 -e ts,js -w src --exec node -r ./loadEnv.js -r ts-node/register src/",
```
[the remark this answers is not in the document]
Comment (gnorbsl) [reply package.json:8]: I tried, it'll be overwritten by my local work setup for whatever reason

[the remark this answers is not in the document]
Comment (Enubia) [reply package.json:8]: hmm ok, well then leave it for now, but also add it to the default.env file please

**Review by Enubia — DISMISSED** [review]

**Review by Enubia — DISMISSED** [review]

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -1,6 +1,6 @@
 name: Run build and test pipeline
 
-on: [push]
+on: [push, pull_request]
```
Comment (Enubia) [inline .github/workflows/build.yml:3]: I removed the pull_request on purpose, if you do that it runs twice in opened PRs

File: prisma/schema.prisma [hunk 7]
```diff
@@ -3,176 +3,205 @@ generator client {
...
+  donationPrice Float
+  status        donation_status @default(DONATED)
+  createdAt     DateTime        @default(now()) @db.Timestamp(6)
+  updatedAt     DateTime        @default(now()) @db.Timestamp(6)
+  deletedAt     DateTime?       @db.Timestamp(6)
+  user          user            @relation(fields: [userId], references: [id])
+  wishcard      wishcard        @relation(fields: [wishcardId], references: [id])
+
+  @@index([userId], name: "donation_userId_idx")
+  @@index([wishcardId], name: "donation_wishcardId_idx")
```
Comment (Jacobjeevan) [inline prisma/schema.prisma:97]: Wishcard to donation is 1:1, so the index here is unnecessary, no? :thinking:

File: prisma/schema.prisma [hunk 8]
```diff
@@ -3,176 +3,205 @@ generator client {
...
+  uid                               String         @unique @db.VarChar
+  role                              roles          @default(DONOR)
+  createdAt                         DateTime       @default(now()) @db.Timestamp(6)
+  updatedAt                         DateTime       @default(now()) @db.Timestamp(6)
+  deletedAt                         DateTime?      @db.Timestamp(6)
+  agency                            agency[]
+  agencyMember                      agencyMember[]
+  donation                          donation[]
+  message                           message[]
+  wishcard_userTowishcard_createdBy wishcard[]     @relation("userTowishcard_createdBy")
```
Comment (Jacobjeevan) [inline prisma/schema.prisma:141]: minor thing, but mind renaming/shortening this to wishcard_createdBy? (Same in line 161).

File: prisma/schema.prisma [hunk 9]
```diff
@@ -3,176 +3,205 @@ generator client {
...
+  agencyId                      Int
+  createdBy                     Int
+  addressId                     Int
+  status                        wishcard_status @default(DRAFT)
+  createdAt                     DateTime        @default(now()) @db.Timestamp(6)
+  updatedAt                     DateTime        @default(now()) @db.Timestamp(6)
+  deletedAt                     DateTime?       @db.Timestamp(6)
+  address                       address         @relation(fields: [addressId], references: [id])
+  agency                        agency          @relation(fields: [agencyId], references: [id])
+  user_userTowishcard_createdBy user            @relation("userTowishcard_createdBy", fields: [createdBy], references: [id])
```
Comment (Jacobjeevan) [inline prisma/schema.prisma:161]: minor thing, but mind renaming/shortening this to wishcard_createdBy? (Same in line 124).

Edit: Hmm, I suggest removing this field (keep it similar to what we had earlier - wish card being related only to the agency and not directly to the user).

We want to display/make cards editable to users of a particular agency anyways.

Comment (Enubia) [reply prisma/schema.prisma:161]: this field is auto-generated by prisma, we don't use that anywhere right now

File: prisma/schema.prisma (hunk 7 above)
Comment (Enubia) [reply prisma/schema.prisma:97]: @gnorbsl

**Pull request merged**